Fix Android project issues

DRSchlaubi opened the pull request:
This PR will try to fix the problems with the android project
Okay it took hours to solve it but it's finally done
Previously the Android project was a subproject of the Quillyjumper project.
This PR uses a different approach, it makes the Android project it's completely own Gradle project and therefor IntelliJ does no longer import it as an android-Gradle project and makes everything an android project
In order to make the project visible in the project tree, I imported it as an external module, that's why the project files are now checked into VCS

DRSchlaubi committed:
Prepare new Android project
- Delete existing android
- Bump gradle version to 6.0.1
- Move assets folder out of Android folder as it was never supposed to be there

DRSchlaubi committed:
Fix Assets issues
- Make "assets" folder a resource of the core project so platform project can inherit it
- Target java 8 on desktop instead of java 7 since 7 reached it's EOL in 2015
also if your Kotlin JVM is 1.8 why target java 7?

DRSchlaubi commented:
This will also Fix #10

DRSchlaubi commented:
Android project is still on Gradle 5 because the Kotlin Android plugin does currently not work with Gradle 6

DRSchlaubi committed:
Add Android Project
- Migrate desktop gradle to kts script (Fix #10)
- Make the Android project indepentent and add the core project as a subproject to avoid previous issues
- Bump gradle to 6.0.1
- Convert buildscript using gradle files to plugins dsl as its newer and more cool

DRSchlaubi committed:
Update .gitignore
- Do no longer ignore .idea and .iml
- Add ignored files

DRSchlaubi committed:
Merge branches 'feature/android-version' and 'master' of https://github.com/Quillraven/QuillyJumper into feature/android-version

 Conflicts:
	android/build.gradle.kts
	build.gradle.kts
	core/build.gradle.kts
	desktop/build.gradle.kts
	settings.gradle.kts

DRSchlaubi committed:
Fix Gradle project
- Remove .gradle files
- Fix buildSrc project

DRSchlaubi committed:
Add .iml for buildSrc

DRSchlaubi committed:
Add Versions.kt again

DRSchlaubi committed:
Merge remote-tracking branch 'origin/master' into feature/android-version

# Conflicts:
#	android/build.gradle.kts
#	build.gradle.kts
#	core/build.gradle.kts
#	desktop/build.gradle.kts
#	settings.gradle.kts

DRSchlaubi committed:
Ad project files

Quillraven commented:
Thanks for the contribution. Can we check that together next week please?

Is it really necessary to commit local files like .idea folder/.iml files? 
Also, seems like some jars and binary files got committed too which should ideally not be part of a repository in my opinion. 

I tried something yesterday - not as clean as your gradle files though - which is now part of the master. I want to take over your PR during the stream step by step to see if at one point something gets messed up again.


Anyway, thanks for the contribution and please cross check my first concerns.

will do some review on specific files soon.

Quillraven commented on .idea/checkstyle-idea.xml:
```diff
@@ -0,0 +1,16 @@
+      </map>
+    </option>
+  </component>
+</project>
```
Is a `.idea` folder really necessary? Normally that is an automatically created folder by Intellij and does not need to be part of the repo?

Also, missing newline at the end. This is missing in multiple files. It can be enabled in Intellij so that it automatically adds a newline if missing.

Quillraven commented on android/.gitignore:
```diff
@@ -0,0 +1,137 @@
+### Gradle Patch ###
+**/build/
+
+# End of https://www.gitignore.io/api/gradle,kotlin,intellij
```
Is it good practice to have  multiple `.gitignore` files? We already have one in the root project. Why not add the new things there?

Quillraven commented on android/.gitignore:
```diff
@@ -0,0 +1,137 @@
+### Gradle Patch ###
+**/build/
+
+# End of https://www.gitignore.io/api/gradle,kotlin,intellij
```
Is it good practice to have multiple .gitignore files? We already have one in the root project. Why not add the new things there?

DRSchlaubi replied:
I've seen this in bigger repos already

Quillraven commented on .idea/checkstyle-idea.xml:
```diff
@@ -0,0 +1,16 @@
+      </map>
+    </option>
+  </component>
+</project>
```
Is a `.idea` folder really necessary? Normally that is an automatically created folder by Intellij and does not need to be part of the repo?

Also, missing newline at the end. This is missing in multiple files. It can be enabled in Intellij so that it automatically adds a newline if missing.

DRSchlaubi replied:
https://intellij-support.jetbrains.com/hc/en-us/articles/206544839
The idea folder is there to share configurations for different user if you don't use it the current setup would not work anymore

Quillraven commented on android/android.iml:
```diff
@@ -0,0 +1,28 @@
+      </configuration>
+    </facet>
+  </component>
+</module>
```
`.iml` file really needed in repository? Why is the project not imported via the `build.gradle.kts` file?

DRSchlaubi replied:
Because we have 2 Gradle project it's easier to import just import one file instead of two. Even eclipse should be able to import .iml files

Quillraven commented on android/build.gradle.kts:
```diff
@@ -1,132 +1,107 @@
 }
 
 android {
-    buildToolsVersion("29.0.2")
```
is `buildToolsVersion` not important?

DRSchlaubi replied:
I have no idea but Android studio did not complain about it so it seems fine

Quillraven commented on android/build.gradle.kts:
```diff
@@ -1,132 +1,107 @@
-            jniLibs.srcDirs("libs")
+            java.srcDir("src/main/kotlin")
+            assets.srcDir("../assets")
+            jniLibs.srcDir("libs")
```
What happens if `aidl.srcDirs`, `renderscript.srcsDirs` and `res.srcDirs` are not linked to `src/main/kotlin`? Is this an issue?

Seems like res files e.g. got moved to `src/main/kotlin/res` - is that ok?

DRSchlaubi replied:
Res should be in `src/main/res` but isn't src the default src folder?

Quillraven commented on android/build.gradle.kts:
```diff
@@ -1,132 +1,107 @@
-
-    }
-    packagingOptions {
-        exclude("META-INF/robovm/ios/robovm.xml")
```
`packagingOptions` not needed?

DRSchlaubi replied:
Dude there is an `ios` in that line so it needs to be remove

Quillraven commented on android/build.gradle.kts:
```diff
@@ -1,132 +1,107 @@
     }
-}
 
-task("run") {
```
What happened to the run task? Is it not needed?

DRSchlaubi replied:
You would only need that when using notepad++ and who TF uses that?
That's what an IDE is for just click the run button in there instead

Quillraven commented on android/buildSrc/build.gradle.kts:
```diff
@@ -0,0 +1,7 @@
+
+plugins {
+    `kotlin-dsl`
+}
```
the `buildSrc` seems to be an experimental feature? I actually like it but I am not sure if it messes up things at the moment.

Will try that again next week during the stream and if possible replace the `rootProject.extra...` parts with the `Versions.kt` again

DRSchlaubi replied:
I also don't like the buildSrc thing probably I will replace it

Quillraven commented on android/buildSrc/buildSrc.iml:
```diff
@@ -0,0 +1,18 @@
+      </configuration>
+    </facet>
+  </component>
+</module>
```
.iml file really part of repository? missing newline at the end

DRSchlaubi replied:
The file is generated so I am not going to at the new line myself

Quillraven commented on android/buildSrc/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,5 @@
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-bin.zip
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
```
another gradle wrapper for the buildSrc directory? Why can we not use the gradle wrapper from root?

Quillraven commented on android/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,5 @@
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-all.zip
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
```
and again  another gradle wrapper for android? same question as for `buildSrc` - can we use the one from root project?

DRSchlaubi replied:
Well no for 2 reasons
- root project uses Gradle 6 and kotlin android does not like Gradle 6
- the android project is separate from the main project that way we get around intellij turning everything into an Android project

Quillraven commented on android/local.properties:
```diff
@@ -0,0 +1,8 @@
+# For customization when using a Version Control System, please read the
+# header note.
+#Thu Dec 26 21:27:41 CET 2019
+sdk.dir=E\:\\WORK\\Android\\SDK
```
`local.properties` should not be part of repo. They are your own local values from what I know

DRSchlaubi replied:
Looks like I forgot to add the Android template to the gitignore

Quillraven commented on android/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,24 @@
+    </activity>
+  </application>
+
+</manifest>
```
Seems like file got moved from `android` to `android source folder`. Is this really ok?

DRSchlaubi replied:
That's how the flutter generator and flutter is by Google so I trust it
Also some example projects did it like this

Quillraven commented on android/src/main/kotlin/com/game/quillyjumper/android/Launcher.kt:
```diff
@@ -1,13 +1,13 @@
+import com.game.quillyjumper.Main
 
-class AndroidLauncher : AndroidApplication() {
+class Launcher : AndroidApplication() {
```
keep the name `AndroidLauncher` because it makes it clearer when searching `Launcher` classes. There is a `DesktopLauncher` for desktop and an `AndroidLauncher` for android.

Quillraven commented on build.gradle.kts:
```diff
@@ -1,41 +1,10 @@
-        classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlinVersion")
-    }
+plugins {
+    kotlin("jvm") version Versions.kotlin apply false // root project does not need Kotlin plugin since it does not contain any source
```
Is this needed in root project as it is anyway specified in `core` and `desktop` project?

DRSchlaubi replied:
It's more clean to have the version only at a single place 
The sub projects only use `kotlin("jvm")` and inherit the version from it's parent

Quillraven commented on desktop/build.gradle.kts:
```diff
@@ -1,54 +1,36 @@
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
 plugins {
+    application
```
what is the purpose of the `application` plugin? Currently I don't have it and it seems to work but I did not try to create  a jar file. Maybe it is needed for that?

DRSchlaubi replied:
The application plugin is there to make your application runnable without it you won't be able to specify the main class so the application won't run

Quillraven reviewed (commented):
Please review comments and let's check together next week.

My main concern at the moment is the multiple gradle wrappers, buildSrc folders, binary files and .idea/.iml files in the repository.

It does not look clean and I prefer the current setup over it.

However, I like gradle files of you a lot more than the current ones because:
- tasks are cleaner
- packaging a jar seems to be easier (shadowJar?)
- buildSrc Versions can be accessed anywhere in the script while rootProject.extra e.g. cannot be accessed in a `plugins` section

DRSchlaubi committed:
Improvements
- Remove buildSrc as it's experimental
- Add Android template to android/.gitignore
- Replace Versions.kt with versions.properties
- Replace Kotlin dependency calls with the kotlin() method

Quillraven commented:
@DRSchlaubi: Thanks a lot for the contribution. I took over most of your adjustments but not the entire thing. Why?

- I prefer having `android` as a subproject treated as a normal module like `core` and `desktop`. That way you can simply import the root `build.gradle.kts` and that's it
- Also, I prefer the `buildSrc` approach since you can use the variables everywhere and you get content assist and type-safety

I adjusted the android `build.gradle.kts` with your changes.
I adjusted the desktop `build.gradle.kts` with your changes including `shadowJar`. I have configured the shadowJar task to create the jar with the correct name and version
I adjusted the  root `build.gradle.kts` with your changes except for the kotlin plugin definition. Seems like it is not necessary there and after all the changes now I was still able to run android and desktop without any issues.

Thanks again for the help - please cross check the master branch and if it is fine for you I will close the issue and PR.

One question that I could not answer regarding the android `build.gradle.kts`:
```kotlin
if (outputDir != null) {
                    copy {
                        from(zipTree(jar))
                        into(outputDir)
                        include("*.so")
                    }
                }
```

When changing that part to the following (which I had):

```diff
if (outputDir != null) {
                    copy {
                        from(zipTree(jar))
                        into(outputDir) 
+                     {
                            include("*.so")
+                     }
                    }
                }
```

I could not deploy on my phone anymore because of a `baseDir cannot be null` problem. Do you know why? What is the difference between `include` outside and inside of the `into` statement?

DRSchlaubi commented:
> I took over most of your adjustments but not the entire thing. Why?

Because that is what you do with good contributions?
> That way you can simply import the root build.gradle.kts and that's it

Yeah but you know what would be better than just importing the build.gradle.kts
A structure that would work out of the box without any configurations the android projects fails "Gradle-syncing" when importing it using IntelliJ's VCS integration https://rice.by.devs-from.asia/idea64_XQPeVRJ6oq.png
My version does not do that it compiles like intended, it also keeps IntelliJ configurations Why would every contributor need to set code style and copyright settings (Well because this project nor has a copyright header neither an actual code style this is not that important)
Also, SUPRISE!?!?!, you still can import the build.gradle.kts file in IntelliJ and any other IDE and even make changes to the android project (in IntelliJ you get Syntax highlighting did not test Eclipse because Eclipse is shit), When importing from .iml IntelliJ honors all project-wide settings and modules and when there is some guy out there who uses IntelliJ and wants to contribute he has a better working project import 
> I adjusted the root build.gradle.kts with your changes except for the Kotlin plugin definition. Seems like it is not necessary there and after all the changes now I was still able to run android and desktop without any issues.

I am sure you considered this but did you also think about why Gradle semi-deprecated the build script DSL and there is a reason why it's not recommended to use? Also adding the plugin to the classpath of the root project
https://github.com/Quillraven/QuillyJumper/blob/05608243c06de2efedaf8588390a31ff72af7a7d/build.gradle.kts#L12-L13
is doing the same thing like plugins did it's just a way of doing it that needs more code
> What is the difference between include outside and inside of the into the statement?

Nothing the `CopySpec#into(Object)` method just returns the `CopySpec` objects you just used and applies the statements to it:

https://github.com/gradle/gradle/blob/b90c4186e8e4901d28379b60e2de1d9caaa554a3/subprojects/core/src/main/java/org/gradle/api/internal/file/copy/CopySpecWrapper.java#L189-L193

> Also, I prefer the buildSrc approach since you can use the variables everywhere and you get content assist and type-safety

Yeah, it's kinda cool to have the type-safety and stuff, but WHY THE HELL DOES IT NEED TO BE A COMPLETELY SEPERATE MODULE!?!? If they would improve that I would consider using it
> I prefer having android as a subproject treated as a normal module like core and desktop

OK, I get that and I prefer that too, but doing that IntelliJ makes everything and Kotlin, Android and Android-Gradle project at the same time, which is not only stupid as hell but also breaks IntelliJ's Gradle integration and some other parts of IntelliJ as IntelliJ does not like doing that (for some reason I don't understand why myself and tried resolving this for ages but it did not work

And now after several hours of work I just get this super confusing Grammer.ly rating:
https://rice.by.devs-from.asia/chrome_51oBdxNWfw.png

> I could not deploy on my phone anymore because of a baseDir cannot be null problem

What are you talking about? I just tried running it and it worked first try (okay it worked second try because my internet cut off during the download of some android stuff, but that's not the fault of the project)

DRSchlaubi commented:
ALSO BRUH https://rice.by.devs-from.asia/gitkraken_dJWdeLEkBN.png
This PR gets replaced by #12
